Update security-secrets-setup README.md (#1836)
edgexfoundry/edgex-go: cmd/security-secrets-setup/README.md

# Testing Instruction for security-secrets-setup

## Executable Build

Use Makefile in the base directory of this repository to build the binary executable of security-secrets-setup:

```sh
make cmd/security-secrets-setup/security-secrets-setup
```

It should create a binary executable under the directory *cmd/security-secrets-setup/* if successfully built.

## Executable Run [generate] subcommand

To run the **generate** subcommand in linux system, change directory to *cmd/security-secrets-setup/* and then run 

```sh

> sudo -E ./security-secrets-setup generate

Author: it needs root privilege to do things

```

> Verify that it runs successfully with generating both Vault and Kong's TLS assets under `/run/edgex/secrets/` directory.

Reviewer: Will also be able to fix some of this when support for `configuration.toml` is added, as /run is really for docker, and it can be in the `res/docker/configuraiton.toml` and the standard `configuration.toml` can just write something to `/tmp/somewhere`

> Verify that it runs successfully with generating both Vault and Kong's TLS assets under `/run/edgex/secrets/` directory.

Author: yea, will need to modify this doc. again when that feature is in.

> Clean up `/run/edgex/secrets` once done verification if necessary.

Author: ok to remove this

## Executable Run [cache] subcommand

Before cache subcommand is issued, make sure cache dir (in this example, it is */tmp/pki/cache*) is non-existing or empty.

To run the **cache** subcommand in linux system, change directory to *cmd/security-secrets-setup/* and then run 

```sh

> mkdir -p /tmp/pki/cache && export PKI_CACHE=/tmp/pki/cache && sudo -E ./security-secrets-setup cache

Reviewer 2: finally these kinds of commands read better in a shell session with one command per line as in:

```suggestion
$ mkdir -p /tmp/pki/cache
$ export PKI_CACHE=/tmp/pki/cache
$ sudo -E ./security-secrets-setup cache
```

```

After successfully run, verify that there are Vault and Kong TLS assets generated and cached into cache dir */tmp/pki/cache* and also deployed from cache directory into the deployed directory */run/edgex/secrets* via `ls` command.

One can also use some file diff tool to compare the TLS files like:

```sh
sudo diff -r /tmp/pki/cache/ca/ca.pem /run/edgex/secrets/ca/ca.pem
```

If compares with the whole parent directories between cache dir and deployed dir, one should only find the difference in generated **sentinel** files.

```sh
sudo diff -r /tmp/pki/cache/ /run/edgex/secrets/

Only in /run/edgex/secrets/ca: .security-secrets-setup.complete
Only in /run/edgex/secrets/edgex-kong: .security-secrets-setup.complete
Only in /run/edgex/secrets/edgex-vault: .security-secrets-setup.complete
```

> Clean up `/run/edgex/secrets` once done verification if necessary.

Reviewer 2: see above comment about clean up comments

## Executable Run [import] subcommand

> To run the **import** subcommand in linux system, change directory to *cmd/security-secrets-setup/* and then run 

Reviewer 2: ```suggestion
To use the **import** subcommand, from the same `cmd/security-secrets-setup` directory run 
```

```sh

> mkdir -p /tmp/pki/cache && export PKI_CACHE=/tmp/pki/cache && sudo -E ./security-secrets-setup import

Reviewer 2: see above comment about having one command per line

```

> There are two cases can happen depends on the status of cache directory:

Reviewer 2: ```suggestion
The **import** subcommand operates differently depending on the status of cache directory:
```

> 1. Cache dir */tmp/pki/cache* is empty:

Reviewer 2: Since there's only 2 possibilities, I think it's simpler to just use plain English and not enumerate them in a number list, like so:

```suggestion
If the cache dir `/tmp/pki/cache` is empty, then the command blocks and waits for ....
```

(also note this now out-of-date per our Slack conversation regarding the blocking...)

Author: i think list reads better, I agree with you on not using the number list but I am ok with changing it to a bullet list not numbered since number list implies the sequence and priority

    In this case, one should expect to see the execution is blocking and waiting for TLS assets to be drop into cache directory before they can be deployed into directory */run/edgex/secrets/*.

2. Cache dir */tmp/pki/cache* has already loaded or cached with generated TLS assets before run import subcommand:

>     One should see the deploying TLS assets happening right away from cached directory */tmp/pki/cache* into deployed directory */run/edgex/secrets*.  One can verify that there are Vault and Kong TLS assets deployed from cache directory */tmp/pki/cache* onto the directory */run/edgex/secrets* via `ls` command.

Reviewer 2: Since this paragraph is just about verifying the files exist I think we should drop it. We already mention above that the files are copied for deployment, so I don't think it's necessary to explain again that files are deployed.

```suggestion
```

Author: i think for the sake of completeness, this should retain

> Clean up `/run/edgex/secrets` once done verification if necessary.

Reviewer 2: see above comment about cleaning things up

## Docker Build

> Use Makefile in the base directory of this repository to build the docker image of security-secrets-setup:

Reviewer 2: ```suggestion
Use the Makefile in the root directory of the repository to build the docker container image for security-secrets-setup:
```

```sh
make docker_security_secrets_setup
```

It should create a docker image with the name `edgexfoundry/docker-edgex-vault:<version>-dev` if sucessfully built.

## Docker Run

To see what the TLS materials are generated inside the docker container, 

> one can run the built docker image above (assuming the version number is `1.1.0`) as:

Reviewer: Docker Run section is still showing the legacy output - should add some text to note that.

Author: Right, that is the current way it is in docker image...this will get changed when the docker image split-ie. after the other issue #1716 is addressed.

Only in /run/edgex/secrets/edgex-vault: .security-secrets-setup.complete
```

## Run **import** subcommand

To use the **import** subcommand, from the same `cmd/security-secrets-setup/` directory run

```sh
 mkdir -p /tmp/pki/cache
 export PKI_CACHE=/tmp/pki/cache
 sudo -E ./security-secrets-setup import
```

The **import** subcommand operates differently depending on the status of cache directory:

- Cache dir `/tmp/pki/cache` is empty:

    In this case, import just errors out: the TLS assets are required to be in the cache directory before they can be deployed into directory `/run/edgex/secrets/`.

- Cache dir `/tmp/pki/cache` has already loaded or cached with generated TLS assets before run **import** subcommand:

    One should see the deploying TLS assets from the cached directory `/tmp/pki/cache` into the deployed directory `/run/edgex/secrets`.  One can verify that there are Vault and Kong TLS assets deployed from cache directory `/tmp/pki/cache` onto the directory `/run/edgex/secrets` via `ls` command.

## Docker Build

Use the Makefile in the root directory of the repository to build the docker container image for `security-secrets-setup`:

```sh

> make docker_security_secrets_setup

Reviewer 3: When I get to the 'make docker_security_secrets_setup' step I see this error message:

```$ make docker_security_secrets_setup
docker build \
	-f cmd/security-secrets-setup/Dockerfile \
	--label "git_sha=21750bd69222753f69a7a217e480b07ca6bb6dd6" \
	-t edgexfoundry/docker-edgex-vault:21750bd69222753f69a7a217e480b07ca6bb6dd6 \
	-t edgexfoundry/docker-edgex-vault:1.1.0-dev \
	.
error checking context: 'can't stat '/home/tpm/go-jim/edgex-go/cmd/security-secrets-setup/logs/security-secrets-setup.log''.
Makefile:174: recipe for target 'docker_security_secrets_setup' failed
make: *** [docker_security_secrets_setup] Error 1
```

Looking at the directory structure of the secrets_setup I see the log file are owned by 'root':
```$ ls -la /home/tpm/go-jim/edgex-go/cmd/security-secrets-setup
-rwxrwxr-x  1 tpm  tpm      3012 Oct  2 10:28 entrypoint.sh
drwxr--r--  2 root root     4096 Oct  2 10:30 logs
-rw-rw-r--  1 tpm  tpm      4514 Oct  2 10:28 main.go
```

Author: yeap, please go to the `root` directory- ie. `edgex-go/` to run the docker build: `make docker_security_secrets_setup` . that way it should be fine. I'll mention that in the document about the root directory you should be run the make docker

mkdir -p /tmp/pki/cache && export PKI_CACHE=/tmp/pki/cache && sudo -E ./security-secrets-setup import
```

There are two cases can happen depending on the status of cache directory:

1. Cache dir */tmp/pki/cache* is empty:

>     In this case, import just errors out: the TLS assets is required to be in the cache directory before they can be deployed into directory */run/edgex/secrets/*.

Reviewer: The intent here was that if the assets are being copied from an image into a mounted docker volume, that the container would wait for this to finish.  Does docker guarantee that this will happen before the container starts executing?
